From 0c7e534e778d62cab4198630f480546856ccabbb Mon Sep 17 00:00:00 2001
From: Pranoy Goru <gorumanip@ami.com>
Date: Mon, 6 May 2024 17:12:05 +0530
Subject: [PATCH 265/265] Added SMBIOS port info and changes in property
 datatype

Description :

Added SMBIOS information in ports section and datatypes of some properties are changed in backend.

Test Case :

1) https://{{BMCIP}}/redfish/v1/Systems/system/Storage/

Response:

{
    "@odata.id": "/redfish/v1/Systems/system/Storage",
    "@odata.type": "#StorageCollection.StorageCollection",
    "Members": [
        {
            "@odata.id": "/redfish/v1/Systems/system/Storage/Nvme"
        }
    ],
    "Members@odata.count": 1,
    "Name": "Storage Collection"
}

2) Get https://{{BMCIP}}/redfish/v1/Systems/system/Storage/Nvme/

Response:

{
    "@odata.id": "/redfish/v1/Systems/system/Storage/Nvme",
    "@odata.type": "#Storage.v1_14_0.Storage",
    "Controllers": {
        "@odata.id": "/redfish/v1/Systems/system/Storage/Nvme/Controllers"
    },
    "Drives": [
        {
            "@odata.id": "/redfish/v1/Systems/system/Storage/Nvme/Drives/0"
        }
    ],
    "Drives@odata.count": 1,
    "Id": "Nvme",
    "Name": "Storage"
}

3) Get https://{{BMCIP}}/redfish/v1/Systems/system/Storage/Nvme/Drives/0

Response:

{
    "@odata.id": "/redfish/v1/Systems/system/Storage/Nvme/Drives/0",
    "@odata.type": "#Drive.v1_7_0.Drive",
    "Id": "0",
    "Links": {
        "Chassis": {
            "@odata.id": "/redfish/v1/Chassis/AC_Baseboard"
        }
    },
    "Manufacturer": "INTEL",
    "Model": "P5800X",
    "Name": "0",
    "Oem": {
        "OpenBmc": {
            "@odata.id": "/redfish/v1/Systems/system/Storage/Nvme/Drives/0#/Oem/OpenBmc",
            "@odata.type": "#OemDrive.v1_0_0.Drive",
            "DeviceLocation": "PCIe Slot 3",
            "Firmware_Version": ""
        }
    },
    "PartNumber": "INTEL SSDPF21Q400GB",
    "SerialNumber": "PHAL14210009400AGN"
}

4) Get https://{{BMCIP}}/redfish/v1/Systems/system/Storage/Nvme/Controllers

Reponse:

{
    "@odata.id": "/redfish/v1/Systems/system/Storage/Nvme/Controllers",
    "@odata.type": "#StorageControllerCollection.StorageControllerCollection",
    "Description": "Collections of storage controllers",
    "Members": [
        {
            "@odata.id": "/redfish/v1/Systems/system/Storage/Nvme/Controllers/0"
        }
    ],
    "Members@odata.count": 1,
    "Name": "Storage Controllers Collections"
}

5) Get https://{{BMCIP}}/redfish/v1/Systems/system/Storage/Nvme/Controllers/0

Response:

{
    "@odata.id": "/redfish/v1/Systems/system/Storage/Nvme/Controllers/0",
    "@odata.type": "#StorageController.v1_6_0.StorageController",
    "Id": "0",
    "Manufacturer": "INTEL",
    "Model": "P5800X",
    "NVMeControllerProperties": {
        "NVMeSMARTCriticalWarnings": {
            "MediaInReadOnly": false,
            "OverallSubsystemDegraded": false,
            "PowerBackupFailed": false,
            "SpareCapacityWornOut": false
        }
    },
    "Name": "NVME controller",
    "Oem": {
        "@odata.type": "#OemAMINVMeMIController.Oem",
        "Ami": {
            "@odata.type": "#OemAMINVMeMIController.Ami",
            "IdenNameSpaceDataStruct": {
                "NamespaceCapacity": 781422768,
                "NamespaceSize": 781422768,
                "NamespaceUtilization": 781422768
            },
            "IdenNameSpaceLBAFmts": {
                "LBAdataSizeofLBAFmtsNo0": 9,
                "LBAdataSizeofLBAFmtsNo1": 9,
                "MetadataSizeofLBAFmtsNo0": 0,
                "MetadataSizeofLBAFmtsNo1": 8,
                "RelativePerformofLBAFmtsNo0": 2,
                "RelativePerformofLBAFmtsNo1": 2
            },
            "IdenNameSpaceList": {
                "NameSpaceListNo0": 1,
                "NameSpaceListNo1": 0
            },
            "IdenPwrStDes": {
                "MaxPowerScaleNo0": 0.01,
                "MaxPowerScaleNo1": 0.01,
                "MaxPowerStateNo0": 1300,
                "MaxPowerStateNo1": 1300
            },
            "NVMeMISubSystemInformation": {
                "@odata.type": "#OemAMINVMeMIController.NVMeMISubSystemInformation",
                "SubSystemHealthStatusPoll": {
                    "CompositeControllerStatus": {
                        "AvailableSpare": 0,
                        "CompositeTemperatureChange": 1,
                        "ControllerEnableChangeOccured": 0,
                        "ControllerFatalStatus": 0,
                        "ControllerStatusChange": 0,
                        "CriticalWarnings": 0,
                        "FirmwareActivated": 0,
                        "NVMSubSystemResetOccured": 0,
                        "NameSpaceAttributeChange": 0,
                        "PercentageUsed": 0,
                        "Ready": 0,
                        "ShutDownStatus": 0
                    },
                    "CompositeTemperature": 28,
                    "DriveFunctional": 1,
                    "PDLU": 0,
                    "Port0PCIeLinkActive": 0,
                    "Port1PCIeLinkActive": 0,
                    "ResetNotRequired": 1,
                    "SmartWarnings": {
                        "UnderThresh": 0
                    }
                },
                "SubSystemInformation": {
                    "MajorVersion": 1,
                    "MinorVersion": 1,
                    "NoOfPorts": 2
                }
            },
            "VitalProductData": {
                "@odata.type": "#OemAMINVMeMIController.VitalProductData",
                "MultiRecordInformation": {
                    "NVMeMultiRecord": {
                        "EPTFactor": 17,
                        "InitialPowerSupplyReq1.8v": "Power supply voltage is not used.",
                        "InitialPowerSupplyReq12v": "Power supply voltage is not used.",
                        "InitialPowerSupplyReq3.3v": "Power supply voltage is not used.",
                        "InitialPowerSupplyReq5v": "Power supply voltage is not used.",
                        "MaxAuxPowerSupplyReq3.3v": "Power supply voltage is not used.",
                        "MaxPowerSupplyReq1.8v": "Power supply voltage is not used.",
                        "MaxPowerSupplyReq12v": "Power supply voltage is not used.",
                        "MaxPowerSupplyReq3.3v": "Power supply voltage is not used.",
                        "MaxPowerSupplyReq5v": "Power supply voltage is not used.",
                        "MaxThermalLoad": "0 Watts",
                        "RecordAreaVersion": 0
                    },
                    "NVMePCIePortMultiRecord": {
                        "MctpSupport": "MCTP support",
                        "PCIeMaximumLinkWidth": "PCIe x4 ",
                        "PCIePortInformation": "PCIe ports have same capabilities",
                        "PCIePortNumber": 0,
                        "PCIeRecordVersion": 130,
                        "PCIeSupportedlinkSpeeds": "Support 2.5GT/s 5.0GT/s 8.0GT/s 16.0GT/s ",
                        "RefClockCapability": "PCIe link supports common RefClk"
                    }
                },
                "ProductInformation": {
                    "IPMIVersion": 1
                }
            }
        }
    },
    "PartNumber": "INTEL SSDPF21Q400GB",
    "Ports": {
        "@odata.id": "/redfish/v1/Systems/system/Storage/Nvme/Controllers/0/Ports"
    },
    "SerialNumber": "PHAL14210009400AGN",
    "Status": {
        "Health": "Warning",
        "HealthRollup": "Critical",
        "State": "Enabled"
    }
}

6) Get https://{{BMCIP}}/redfish/v1/Systems/system/Storage/Nvme/Controllers/0/Ports

Response:

{
    "@odata.id": "/redfish/v1/Systems/system/Storage/Nvme/Controllers/0/Ports",
    "@odata.type": "#PortCollection.PortCollection",
    "Description": "The Collections of Ports",
    "Members": [
        {
            "@odata.id": "/redfish/v1/Systems/system/Storage/Nvme/Controllers/0/Ports/0"
        },
        {
            "@odata.id": "/redfish/v1/Systems/system/Storage/Nvme/Controllers/0/Ports/2"
        }
    ],
    "Members@odata.count": 2,
    "Name": "Port Collections"
}

7) Get https://{{BMCIP}}/redfish/v1/Systems/system/Storage/Nvme/Controllers/0/Ports/0

Response:

{
    "@odata.id": "/redfish/v1/Systems/system/Storage/Nvme/Controllers/0/Ports/0",
    "@odata.type": "#Port.v1_3_0.Port",
    "Id": "0",
    "Name": "PCIe",
    "Oem": {
        "Ami": {
            "@odata.type": "#AMINVMeMIPort.Ami",
            "PortInformation": {
                "@odata.type": "#AMINVMeMIPort.PortInformation",
                "CurrentLinkSpeed": "InActive",
                "MaxLinkWidth": "PCIe x4",
                "MaxMCTPTransUnitSize": 0,
                "MaxPayloadSize": "128 bytes",
                "NegotiatedLinkWidth": "InActive",
                "PortNo": "0",
                "PortStatus": "Active",
                "PortType": "PCIe",
                "SupportedLinkSpeeds": "2.5 GT/s, 5.0 GT/s, 8.0 GT/s, 16.0 GT/s"
            }
        }
    }
}

8) Get https://{{BMCIP}}/redfish/v1/Systems/system/Storage/Nvme/Controllers/0/Ports/2

Response:

{
    "@odata.id": "/redfish/v1/Systems/system/Storage/Nvme/Controllers/0/Ports/2",
    "@odata.type": "#Port.v1_3_0.Port",
    "Id": "2",
    "Name": "SMBus",
    "Oem": {
        "Ami": {
            "@odata.type": "#AMINVMeMIPort.Ami",
            "PortInformation": {
                "@odata.type": "#AMINVMeMIPort.PortInformation",
                "CurrentAddress": 83,
                "EndPointAddress": 29,
                "MaxEndPointFrequency": "400kHz",
                "MaxFrequency": "400kHz",
                "MaxMCTPTransUnitSize": 240,
                "NVMeBasicManagementCommand": "Yes",
                "PortNo": "2",
                "PortStatus": "Active",
                "PortType": "SMBus"
            }
        }
    }
}

Signed-off-by: Pranoy Goru <gorumanip@ami.com>
---
 redfish-core/lib/storage_nvme.hpp | 269 ++++++++++++++++++------------
 1 file changed, 165 insertions(+), 104 deletions(-)

diff --git a/redfish-core/lib/storage_nvme.hpp b/redfish-core/lib/storage_nvme.hpp
index dfaea9b3..e4f06f03 100644
--- a/redfish-core/lib/storage_nvme.hpp
+++ b/redfish-core/lib/storage_nvme.hpp
@@ -623,16 +623,16 @@ inline void getVPDNVMeMultiRecordInfo(
             return;
         }
         const uint8_t* eptFactor = nullptr;
-        const uint8_t* initialPowerSupplyReq18v = nullptr;
-        const uint8_t* initialPowerSupplyReq12v = nullptr;
-        const uint8_t* initialPowerSupplyReq33v = nullptr;
-        const uint8_t* initialPowerSupplyReq5v = nullptr;
-        const uint8_t* maxAuxPowerSupplyReq33v = nullptr;
-        const uint8_t* maxPowerSupplyReq18v = nullptr;
-        const uint8_t* maxPowerSupplyReq12v = nullptr;
-        const uint8_t* maxPowerSupplyReq33v = nullptr;
-        const uint8_t* maxPowerSupplyReq5v = nullptr;
-        const uint8_t* maxThermalLoad = nullptr;
+        const std::string* initialPowerSupplyReq18v = nullptr;
+        const std::string* initialPowerSupplyReq12v = nullptr;
+        const std::string* initialPowerSupplyReq33v = nullptr;
+        const std::string* initialPowerSupplyReq5v = nullptr;
+        const std::string* maxAuxPowerSupplyReq33v = nullptr;
+        const std::string* maxPowerSupplyReq18v = nullptr;
+        const std::string* maxPowerSupplyReq12v = nullptr;
+        const std::string* maxPowerSupplyReq33v = nullptr;
+        const std::string* maxPowerSupplyReq5v = nullptr;
+        const std::string* maxThermalLoad = nullptr;
         const uint8_t* recordAreaVersion = nullptr;
         //      const std::vector<uint8_t*> totalNVMCapacity = nullptr;
 
@@ -771,13 +771,13 @@ inline void getVPDNVMePCIePortRecord(
             // this interface isn't necessary
             return;
         }
-        const uint8_t* mctpSupport = nullptr;
-        const uint8_t* pcieMaximumLinkWidth = nullptr;
-        const uint8_t* pciePortInformation = nullptr;
+        const std::string* mctpSupport = nullptr;
+        const std::string* pcieMaximumLinkWidth = nullptr;
+        const std::string* pciePortInformation = nullptr;
         const uint8_t* pciePortNumber = nullptr;
         const uint8_t* pcieRecordVersion = nullptr;
-        const uint8_t* pcieSupportedlinkSpeeds = nullptr;
-        const uint8_t* refClockCapability = nullptr;
+        const std::string* pcieSupportedlinkSpeeds = nullptr;
+        const std::string* refClockCapability = nullptr;
 
         const bool success = sdbusplus::unpackPropertiesNoThrow(
             dbus_utils::UnpackErrorPrinter(), propertiesList, "mctpSup",
@@ -799,14 +799,10 @@ inline void getVPDNVMePCIePortRecord(
         }
         if (pcieMaximumLinkWidth != nullptr)
         {
-            if (linkWidth.find(*pcieMaximumLinkWidth) != linkWidth.end())
-            {
-                asyncResp->res.jsonValue["Oem"]["Ami"]["VitalProductData"]
-                                        ["MultiRecordInformation"]
-                                        ["NVMePCIePortMultiRecord"]
-                                        ["PCIeMaximumLinkWidth"] =
-                    linkWidth.find(*pcieMaximumLinkWidth)->second;
-            }
+            asyncResp->res
+                .jsonValue["Oem"]["Ami"]["VitalProductData"]
+                          ["MultiRecordInformation"]["NVMePCIePortMultiRecord"]
+                          ["PCIeMaximumLinkWidth"] = *pcieMaximumLinkWidth;
         }
         if (pciePortInformation != nullptr)
         {
@@ -831,33 +827,25 @@ inline void getVPDNVMePCIePortRecord(
         }
         if (pcieSupportedlinkSpeeds != nullptr)
         {
-            if (linkSpeed.find(*pcieSupportedlinkSpeeds) != linkSpeed.end())
-            {
-                asyncResp->res.jsonValue["Oem"]["Ami"]["VitalProductData"]
-                                        ["MultiRecordInformation"]
-                                        ["NVMePCIePortMultiRecord"]
-                                        ["PCIeSupportedlinkSpeeds"] =
-                    linkSpeed.find(*pcieSupportedlinkSpeeds)->second;
-            }
-            else
-            {
-                asyncResp->res.jsonValue["Oem"]["Ami"]["VitalProductData"]
-                                        ["MultiRecordInformation"]
-                                        ["NVMePCIePortMultiRecord"]
-                                        ["PCIeSupportedlinkSpeeds"] =
-                    "Not Supported";
-            }
+            asyncResp->res
+                .jsonValue["Oem"]["Ami"]["VitalProductData"]
+                          ["MultiRecordInformation"]["NVMePCIePortMultiRecord"]
+                          ["PCIeSupportedlinkSpeeds"] =
+                *pcieSupportedlinkSpeeds;
+        }
+        else
+        {
+            asyncResp->res
+                .jsonValue["Oem"]["Ami"]["VitalProductData"]
+                          ["MultiRecordInformation"]["NVMePCIePortMultiRecord"]
+                          ["PCIeSupportedlinkSpeeds"] = "Not Supported";
         }
         if (refClockCapability != nullptr)
         {
-            if (refClock.find(*refClockCapability) != refClock.end())
-            {
-                asyncResp->res.jsonValue["Oem"]["Ami"]["VitalProductData"]
-                                        ["MultiRecordInformation"]
-                                        ["NVMePCIePortMultiRecord"]
-                                        ["RefClockCapability"] =
-                    refClock.find(*refClockCapability)->second;
-            }
+            asyncResp->res
+                .jsonValue["Oem"]["Ami"]["VitalProductData"]
+                          ["MultiRecordInformation"]["NVMePCIePortMultiRecord"]
+                          ["RefClockCapability"] = *refClockCapability;
         }
     });
 }
@@ -1167,36 +1155,6 @@ inline void requestRoutesNvmeControllersInstance(App& app)
     });
 }
 
-inline void getNvmePorts(const std::shared_ptr<bmcweb::AsyncResp>& asyncResp,
-                         const std::string& path)
-{
-    nlohmann::json& controllerArray = asyncResp->res.jsonValue["Members"];
-    controllerArray = nlohmann::json::array();
-    nlohmann::json& count = asyncResp->res.jsonValue["Members@odata.count"];
-    count = 0;
-
-    sdbusplus::asio::getProperty<uint8_t>(
-        *crow::connections::systemBus, nvmeServiceName, path,
-        "xyz.openbmc_project.Common.SubInfoData", "numOfPorts",
-        [asyncResp, &controllerArray,
-         &count](const boost::system::error_code& ec, const uint8_t ports) {
-        if (ec)
-        {
-            return;
-        }
-        for (int var = 0; var < ports; var++)
-        {
-            nlohmann::json::object_t controllerJson;
-            const std::string str =
-                "/redfish/v1/Systems/system/Storage/Nvme/Controllers/0/Ports/" +
-                std::to_string(var);
-            controllerJson["@odata.id"] = str;
-            controllerArray.push_back(std::move(controllerJson));
-        }
-        count = controllerArray.size();
-    });
-}
-
 inline void requestRoutesNvmePorts(App& app)
 {
     BMCWEB_ROUTE(
@@ -1234,7 +1192,15 @@ inline void requestRoutesNvmePorts(App& app)
                                            controllerId);
                 return;
             }
-            const std::string& path = controller->first;
+            const dbus::utility::MapperServiceMap& connectionNames =
+                controller->second;
+            if (connectionNames.size() != 1)
+            {
+                BMCWEB_LOG_ERROR("Connection size ", connectionNames.size(),
+                                 "not equal to 1");
+                messages::internalError(asyncResp->res);
+                return;
+            }
             asyncResp->res.jsonValue["@odata.type"] =
                 "#PortCollection.PortCollection";
             asyncResp->res.jsonValue["@odata.id"] =
@@ -1243,7 +1209,29 @@ inline void requestRoutesNvmePorts(App& app)
             asyncResp->res.jsonValue["Name"] = "Port Collections";
             asyncResp->res.jsonValue["Description"] =
                 "The Collections of Ports";
-            getNvmePorts(asyncResp, path);
+            nlohmann::json& controllerArray =
+                asyncResp->res.jsonValue["Members"];
+            controllerArray = nlohmann::json::array();
+            nlohmann::json& count =
+                asyncResp->res.jsonValue["Members@odata.count"];
+            count = 0;
+            std::string key("xyz.openbmc_project.Common.PortInfoData_Port");
+
+            for (const std::string& interface : connectionNames[0].second)
+            {
+                std::size_t lastPos = interface.rfind(key);
+                if (lastPos == std::string::npos ||
+                    (interface.size() <= lastPos + 1))
+                {
+                    BMCWEB_LOG_ERROR("Failed to find", key, " in ", interface);
+                    continue;
+                }
+                controllerArray.push_back(
+                    {{"@odata.id",
+                      "/redfish/v1/Systems/system/Storage/Nvme/Controllers/0/Ports/" +
+                          interface.substr(lastPos + key.length())}});
+            }
+            count = controllerArray.size();
         });
     });
 }
@@ -1260,11 +1248,8 @@ inline void getInactiveInfo(const std::shared_ptr<bmcweb::AsyncResp>& asyncResp)
 inline void getPCIeInfo(const std::shared_ptr<bmcweb::AsyncResp>& asyncResp,
                         const std::string& interface, const std::string& path)
 {
-    asyncResp->res.jsonValue["Name"] = "PCIe";
     asyncResp->res.jsonValue["Oem"]["Ami"]["PortInformation"]["PortStatus"] =
         "Active";
-    asyncResp->res.jsonValue["Oem"]["Ami"]["PortInformation"]["PortType"] =
-        "PCIe";
     sdbusplus::asio::getAllProperties(
         *crow::connections::systemBus, nvmeServiceName, path, interface,
         [asyncResp](const boost::system::error_code& ec,
@@ -1276,19 +1261,20 @@ inline void getPCIeInfo(const std::shared_ptr<bmcweb::AsyncResp>& asyncResp,
             // this interface isn't necessary
             return;
         }
-        const int32_t* maxMCTPTransUnitSize = nullptr;
+        const uint16_t* maxMCTPTransUnitSize = nullptr;
         const uint8_t* currentLinkSpeed = nullptr;
         const uint8_t* supportedLinkSpeeds = nullptr;
         const uint8_t* maxPayloadSize = nullptr;
         const uint8_t* maxLinkWidth = nullptr;
         const uint8_t* negotiatedLinkWidth = nullptr;
+        const std::string* portType = nullptr;
 
         const bool success = sdbusplus::unpackPropertiesNoThrow(
-            dbus_utils::UnpackErrorPrinter(), propertiesList, "maxPayloadSize",
-            maxPayloadSize, "maxMCTPTranUnitSize", maxMCTPTransUnitSize,
-            "maxLinkWidth", maxLinkWidth, "negLinkWidth", negotiatedLinkWidth,
-            "supLinkSpeedVector", supportedLinkSpeeds, "curLinkSpeed",
-            currentLinkSpeed);
+            dbus_utils::UnpackErrorPrinter(), propertiesList, "MaxPayloadSize",
+            maxPayloadSize, "MaxMCTPTranUnitSize", maxMCTPTransUnitSize,
+            "MaxLinkWidth", maxLinkWidth, "NegLinkWidth", negotiatedLinkWidth,
+            "SupLinkSpeedVector", supportedLinkSpeeds, "CurLinkSpeed",
+            currentLinkSpeed, "PortType", portType);
 
         if (!success)
         {
@@ -1357,6 +1343,91 @@ inline void getPCIeInfo(const std::shared_ptr<bmcweb::AsyncResp>& asyncResp,
                                         ["CurrentLinkSpeed"] = "Not Supported";
             }
         }
+        if (portType != nullptr)
+        {
+            asyncResp->res.jsonValue["Name"] = *portType;
+            asyncResp->res.jsonValue["Oem"]["Ami"]["PortInformation"]
+                                    ["PortType"] = *portType;
+        }
+    });
+}
+
+inline void getSMBusInfo(const std::shared_ptr<bmcweb::AsyncResp>& asyncResp,
+                         const std::string& interface, const std::string& path)
+{
+    asyncResp->res.jsonValue["Oem"]["Ami"]["PortInformation"]["PortStatus"] =
+        "Active";
+    sdbusplus::asio::getAllProperties(
+        *crow::connections::systemBus, nvmeServiceName, path, interface,
+        [asyncResp](const boost::system::error_code& ec,
+                    const std::vector<
+                        std::pair<std::string, dbus::utility::DbusVariantType>>&
+                        propertiesList) {
+        if (ec)
+        {
+            // this interface isn't necessary
+            return;
+        }
+        const uint8_t* curVPDSMBusAddr = nullptr;
+        const uint8_t* curMgtEndptSMbusAddr = nullptr;
+        const std::string* maxVPDSMbusFreq = nullptr;
+        const std::string* maxMgtEndptSMbusFreq = nullptr;
+        const uint16_t* maxMCTPTranUnitSize = nullptr;
+        const std::string* nvmeBasicMgtCmd = nullptr;
+        const std::string* portType = nullptr;
+
+        const bool success = sdbusplus::unpackPropertiesNoThrow(
+            dbus_utils::UnpackErrorPrinter(), propertiesList, "CurVPDSMBusAddr",
+            curVPDSMBusAddr, "CurMgtEndptSMbusAddr", curMgtEndptSMbusAddr,
+            "MaxVPDSMbusFreq", maxVPDSMbusFreq, "MaxMgtEndptSMbusFreq",
+            maxMgtEndptSMbusFreq, "MaxMCTPTranUnitSize", maxMCTPTranUnitSize,
+            "NVMeBasicMgtCmd", nvmeBasicMgtCmd, "PortType", portType);
+
+        if (!success)
+        {
+            messages::internalError(asyncResp->res);
+            return;
+        }
+
+        if (curVPDSMBusAddr != nullptr)
+        {
+            asyncResp->res.jsonValue["Oem"]["Ami"]["PortInformation"]
+                                    ["CurrentAddress"] = *curVPDSMBusAddr;
+        }
+        if (curMgtEndptSMbusAddr != nullptr)
+        {
+            asyncResp->res.jsonValue["Oem"]["Ami"]["PortInformation"]
+                                    ["EndPointAddress"] = *curMgtEndptSMbusAddr;
+        }
+        if (maxVPDSMbusFreq != nullptr)
+        {
+            asyncResp->res.jsonValue["Oem"]["Ami"]["PortInformation"]
+                                    ["MaxFrequency"] = *maxVPDSMbusFreq;
+        }
+        if (maxMgtEndptSMbusFreq != nullptr)
+        {
+            asyncResp->res.jsonValue["Oem"]["Ami"]["PortInformation"]
+                                    ["MaxEndPointFrequency"] =
+                *maxMgtEndptSMbusFreq;
+        }
+        if (maxMCTPTranUnitSize != nullptr)
+        {
+            asyncResp->res.jsonValue["Oem"]["Ami"]["PortInformation"]
+                                    ["MaxMCTPTransUnitSize"] =
+                *maxMCTPTranUnitSize;
+        }
+        if (nvmeBasicMgtCmd != nullptr)
+        {
+            asyncResp->res.jsonValue["Oem"]["Ami"]["PortInformation"]
+                                    ["NVMeBasicManagementCommand"] =
+                *nvmeBasicMgtCmd;
+        }
+        if (portType != nullptr)
+        {
+            asyncResp->res.jsonValue["Name"] = *portType;
+            asyncResp->res.jsonValue["Oem"]["Ami"]["PortInformation"]
+                                    ["PortType"] = *portType;
+        }
     });
 }
 
@@ -1407,11 +1478,11 @@ inline void requestRoutesNvmePortsInstance(App& app)
             const std::string interface =
                 "xyz.openbmc_project.Common.PortInfoData_Port" + portId;
 
-            sdbusplus::asio::getProperty<uint8_t>(
+            sdbusplus::asio::getProperty<std::string>(
                 *crow::connections::systemBus, nvmeServiceName, path, interface,
-                "portType",
+                "PortType",
                 [asyncResp, interface, portId,
-                 path](const boost::system::error_code& ec2, uint8_t type) {
+                 path](const boost::system::error_code& ec2, std::string type) {
                 if (ec2)
                 {
                     return;
@@ -1423,23 +1494,13 @@ inline void requestRoutesNvmePortsInstance(App& app)
                     "#AMINVMeMIPort.PortInformation";
                 asyncResp->res.jsonValue["Oem"]["Ami"]["PortInformation"]
                                         ["PortNo"] = portId;
-                if (type == 0)
-                {
-                    getInactiveInfo(asyncResp);
-                }
-                else if (type == 1)
+                if (type == "PCIe")
                 {
                     getPCIeInfo(asyncResp, interface, path);
                 }
-                else if (type == 2)
+                else if (type == "SMBus")
                 {
-                    //  getSMBusInfo(asyncResp, interface, path); Still backend
-                    //  support need to add
-                    asyncResp->res.jsonValue["Name"] = "SMBus";
-                    asyncResp->res.jsonValue["Oem"]["Ami"]["PortInformation"]
-                                            ["PortStatus"] = "Active";
-                    asyncResp->res.jsonValue["Oem"]["Ami"]["PortInformation"]
-                                            ["PortType"] = "SMBus";
+                    getSMBusInfo(asyncResp, interface, path);
                 }
             });
         });
-- 
2.34.1

